### PR TITLE
VS Expansion Caching

### PR DESCRIPTION
**Pull request by TahaAttari**
[APHL-1180] Add an interface and implementation for VS Expansion Caching

Corresponding VSM PR: https://github.com/DBCG/aphl-vsm/pull/466

[APHL-1180]: https://alphora.atlassian.net/browse/APHL-1180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by Chris0296 — APPROVED** [review]
Looks good to me!

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/common/PackageProcessor.java [hunk 1]
```diff
@@ -20,7 +20,7 @@ public class PackageProcessor implements IPackageProcessor {
     public PackageProcessor(Repository repository) {
         this.repository = repository;
         this.fhirVersion = this.repository.fhirContext().getVersion().getVersion();
-        packageVisitor = new PackageVisitor(this.repository);
+        packageVisitor = new PackageVisitor(this.repository, null);
```
Comment (barhodes) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/common/PackageProcessor.java:23]: The visitor has a constructor with just the repository.  No need to add all these null parameters in.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java [hunk 2]
```diff
@@ -51,6 +51,17 @@ public PackageVisitor(Repository repository) {
         super(repository);
         terminologyServerClient = new TerminologyServerClient(fhirContext());
         expandHelper = new ExpandHelper(this.repository, terminologyServerClient);
+        setupResourceTypes();
```
Comment (barhodes) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java:54]:
```suggestion
        this(repository, null);
```

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/BaseKnowledgeArtifactVisitor.java [hunk 3]
```diff
@@ -34,11 +34,17 @@
 public abstract class BaseKnowledgeArtifactVisitor implements IKnowledgeArtifactVisitor {
     String isOwnedUrl = "http://hl7.org/fhir/StructureDefinition/artifact-isOwned";
     protected final Repository repository;
+    protected Optional<IValueSetExpansionCache> valueSetExpansionCache = Optional.empty();
 
     protected BaseKnowledgeArtifactVisitor(Repository repository) {
         this.repository = repository;
     }
```
Comment (barhodes) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/BaseKnowledgeArtifactVisitor.java:41]:
```suggestion
    protected final Optional<IValueSetExpansionCache> valueSetExpansionCache;

    protected BaseKnowledgeArtifactVisitor(Repository repository) {
        this.repository = repository;
        this.valueSetExpansionCache = Optional.empty();
    }
```

Comment (barhodes) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/BaseKnowledgeArtifactVisitor.java:41]: Or just get rid of this constructor.  The children can still have a constructor without a cache.

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/visitor/dstu3/PackageVisitorTests.java [hunk 4]
```diff
@@ -59,7 +70,7 @@ void visitLibraryTest() {
         Bundle loadedBundle = (Bundle) jsonParser.parseResource(
                 PackageVisitorTests.class.getResourceAsStream("Bundle-ersd-example-naive.json"));
         repo.transaction(loadedBundle);
-        var packageVisitor = new PackageVisitor(repo);
+        var packageVisitor = new PackageVisitor(repo, null);
```
Comment (barhodes) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/visitor/dstu3/PackageVisitorTests.java:73]: No reason for any of these to change.

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/visitor/dstu3/PackageVisitorTests.java [hunk 5]
```diff
@@ -413,4 +424,42 @@ void packageOperation_should_respect_include() {
...
         }
     }
+
+    @Test
+    public void packageVisitorShouldUseExpansionCacheIfProvided() {
+        // Arrange
+        Bundle bundle = (Bundle) jsonParser.parseResource(
+                PackageVisitorTests.class.getResourceAsStream("Bundle-ersd-small-active.json"));
+        repo.transaction(bundle);
+        Library library = repo.read(Library.class, new IdType("Library/SpecificationLibrary"))
```
Comment (barhodes) [inline cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/visitor/dstu3/PackageVisitorTests.java:434]: Nitpick: I'd declare all of these with var.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java [hunk 6]
```diff
@@ -184,15 +195,34 @@ protected void handleValueSets(IBaseBundle packagedBundle, Optional<IBaseResourc
                 .filter(r -> r.fhirType().equals("ValueSet"))
                 .map(v -> (IValueSetAdapter) createAdapterForResource(v))
                 .collect(Collectors.toList());
-
-        valueSets.stream()
-                .forEach(valueSet -> expandHelper.expandValueSet(
+        var expansionCache = this.getExpansionCache();
```
Comment (barhodes) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java:198]: Nitpick: Don't need this 'this.' here.

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java [hunk 7]
```diff
@@ -184,15 +195,34 @@ protected void handleValueSets(IBaseBundle packagedBundle, Optional<IBaseResourc
...
+                var cachedExpansion = expansionCache
+                        .get()
+                        .getExpansionForCanonical(v.getCanonical(), expansionParamsHash.orElse(null));
+                if (cachedExpansion != null) {
+                    v.setExpansion(cachedExpansion.getExpansion());
+                    expandedList.add(v.getUrl());
+                }
+            });
+        }
+        valueSets.stream().forEach(valueSet -> {
```
Comment (barhodes) [inline cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java:212]: We are iterating over each ValueSet twice here if we have an expansion cache.  This should be refactored to iterate only a single time.

**Review by barhodes — CHANGES_REQUESTED** [review]

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/common/PackageProcessor.java (hunk 1 above)
Comment (TahaAttari) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/common/PackageProcessor.java]: fixed

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java (hunk 7 above)
Comment (barhodes) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java]: I see.  Yeah, that seems reasonable.

File: cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/visitor/dstu3/PackageVisitorTests.java (hunk 4 above)
Comment (TahaAttari) [reply cqf-fhir-cr/src/test/java/org/opencds/cqf/fhir/cr/visitor/dstu3/PackageVisitorTests.java]: fixed

File: cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java (hunk 6 above)
Comment (TahaAttari) [reply cqf-fhir-cr/src/main/java/org/opencds/cqf/fhir/cr/visitor/PackageVisitor.java]: done 👍🏾

**Review by barhodes — APPROVED** [review]

**Pull request merged**